Psipred

sjanssen2 opened the pull request:
A recipe for the program Psipred from Tomasz former lab. I had to add a Perl wrapper to make it more convenient to use.

sjanssen2 committed:
added recipe for psipred 4.01

sjanssen2 committed:
now with message to the user

sjanssen2 committed:
check what is going on in Travis

sjanssen2 committed:
fix

sjanssen2 committed:
fix for fix

sjanssen2 committed:
debug

sjanssen2 committed:
debug..

sjanssen2 committed:
debug message

sjanssen2 committed:
changed to blast 2.2.26 and removed debug for blastpgp

sjanssen2 committed:
better post-link message

sjanssen2 committed:
Merge branch 'master' of https://github.com/biocore/conda-recipes into psipred

sjanssen2 committed:
exchanged histid by uname -n since the former is not available under OS x

sjanssen2 committed:
OSX seems to set TMPDIR to "1" ?!

qiyunzhu commented on recipes/psipred/perlwrapper.patch:
```diff
@@ -0,0 +1,179 @@
++	print "Cleaning up ...\n";
++	my @filesToDelete = ('input', 'psiblast.chk', 'psiblast.blast', 'psiblast.pn', 'psiblast.sn', 'psiblast.mn', 'psiblast.aux', 'pass1.ss');
++	push @filesToDelete, 'psiblast.mtx' if ($mtx != 1);
++	foreach my $file (@filesToDelete) {
```
I don't think we have to pursue slim Perl codes as we do for Python, but in case you are interested, this loop can be written as: `qx(rm -f $tmproot/$_) for @filesToDelete;`

sjanssen2 replied:
I like that one, because it also would only cause one system call, right?

qiyunzhu replied:
Well no, it is but another form of writing for loop. There are still multiple system calls.

qiyunzhu commented on recipes/psipred/perlwrapper.patch:
```diff
@@ -0,0 +1,179 @@
++	print "Pass1 ...\n";
++
++	my $filename = $tmproot.'/'.basename($inputfile);
++	my $cmd = $execdir.'psipred '.$tmproot.'/psiblast.mtx '.$datadir.'weights.dat '.$datadir.'weights.dat2 '.$datadir.'/weights.dat3 > '.$tmproot.'/pass1.ss';
```
An alternative form:
`my $cmd = "${execdir}psipred $tmproot/psiblast.mtx ${datadir}weights.dat..."`
I have the impression that Perl people prefer double quotes to single quotes.

qiyunzhu commented on recipes/psipred/perlwrapper.patch:
```diff
@@ -0,0 +1,179 @@
++				"mtx=i" => \$mtx,
++		        "h"  => sub {&usage;});
++
++	&usage if (!$ARGV[0]);
```
Be careful here. Unlike Python, Perl's `ARGV` starts with 0. That is, in command `perl script.pl infile outfile`. `$ARGV[0]` is `infile`. In your code, if the user doesn't enter any argument, this will return an undefined error.
Maybe you can do `&usage unless @ARGV;`

sjanssen2 replied:
You are almost right. As you wrote $ARGV[0] will be undefined. However, I make use of that and test in the if-clause for being undefined. If so, the usage message is printed.

qiyunzhu replied:
Sounds good!

qiyunzhu commented on recipes/psipred/perlwrapper.patch:
```diff
@@ -0,0 +1,179 @@
++	qx($cmd);
++	my $status = $?;
++	if ($status != 0) {
++		system("tail ${tmproot}/psiblast.blast");
```
Is there a reason you do `qx` in some places and do `system` in others?

sjanssen2 replied:
no. It's copy and pasted from the original authors sources.

qiyunzhu replied:
I see.

qiyunzhu commented on recipes/psipred/perlwrapper.patch:
```diff
@@ -0,0 +1,179 @@
++	($inputfile) = @ARGV;
++	
++	# add a / to the end of dir if it's not empty or already existing
++	foreach my $dir ($ncbidir, $execdir, $datadir) {
```
If you love regex and one-liner: `s/([^\/])$/$1\// for ($ncbidir, $execdir, $datadir);`

sjanssen2 replied:
I will leave as is, because of resemblance to author's code. But I like your style :-)

qiyunzhu replied:
Thank you!

qiyunzhu commented on recipes/psipred/perlwrapper.patch:
```diff
@@ -0,0 +1,179 @@
++	}
++	
++	# check if input file exist
++	if (! -e $inputfile) {
```
Btw, Perl has `unless`.

sjanssen2 replied:
I always find "unless" harder to read than "if not"

qiyunzhu replied:
No problem! Your Perl style is not necessarily the slimmest, but it is easier to understand and maintain. That's also why Perl is replaced by Python.

qiyunzhu commented on recipes/psipred/perlwrapper.patch:
```diff
@@ -0,0 +1,179 @@
++				"d=s" => \$dbname,
++				"n=s" => \$ncbidir,
++				"mtx=i" => \$mtx,
++		        "h"  => sub {&usage;});
```
insufficient indent?

sjanssen2 replied:
ha. You wrote too much Python! It's due to different white spaces. I changed it.

qiyunzhu replied:
:)

qiyunzhu commented on recipes/psipred/perlwrapper.patch:
```diff
@@ -0,0 +1,179 @@
++my $tmproot = ($ENV{TMPDIR} ? exists $ENV{TMPDIR} : '').'psitmp'.$$.$hostid;
++qx(mkdir -p $tmproot && cp -f $inputfile $tmproot/input); #($? != 0)
++
++if ($mtx != 1) {
```
May be helpful to add a validation of the value of `$mtx` (0 or 1).

sjanssen2 replied:
we are lazy in Perl. We assume that the default behaviour is running blast and only if the user read the doc and knows what he does he would set mtx to 1 to change default behaviour. Thus, I don't want to bother the user with incorrectly set values for mtx.

qiyunzhu replied:
Sounds good!

qiyunzhu commented on recipes/psipred/perlwrapper.patch:
```diff
@@ -0,0 +1,179 @@
++	
++	my $pwd = qx(pwd); chomp $pwd; $pwd .= '/';
++	print "Final output files: ";
++	foreach my $file (@{$resfiles}) {
```
This code will leave an extra space in the end of the file list. How about:
```
print join(" ", map { s/^$pwd//r } @{$resfiles});
```

sjanssen2 replied:
cool

qiyunzhu reviewed (commented):
Hello @sjanssen2 I read thru your Perl code. I was impressed! You code is always clean and strict. I made quite a few comments, but most of them (except for line #116) are just suggestions. Even without changes, the code should run smoothly. It is all up to you. I am fine with the code!

sjanssen2 committed:
incorporated Qiyun's comments to my Perl wrapper

qiyunzhu commented:
@sjanssen2 Everything looks good!

sjanssen2 committed:
corrected a syntax error

sjanssen2 commented:
Hey @mortonjt could you please have a look at this PR?

qiyunzhu commented:
Everything looks good. @RNAer can you merge?

tkosciolek commented:
* make `post-link.sh` more verbose, e.g. "PSIPRED 4 installation successful! Type 'runpsipred.pl' to run."

* can we call `runpsipred` `run_psipred` for consistency with disopred and memsat-svm? (I'm just being a pain in the ass with this one)

* would it be possible to add "-r" option to `run_psipred.pl` to display README?

* `run_psipred.pl` with default parameters puts output in `psitmp*` folder, not the folder it was executed from

* there are no `.mtx` files left after `run_psipred.pl` run, only `.horiz` and `.ss2`

sjanssen2 commented:
+ I changed the message
+ wrapper is now renamed into run_psipred.pl
+ I added the option -o to specify an output directory. By default it will be ./psitmpXXX
+ the *.mtx is now part of the regular output.

- I did not add the ability to print the readme file, since it is quit unusual to do something like this. It would be more appropriate to convert it into a man file, but I don't know how to do it.

P.S. do you have further comments on disopred and memsat-svm?

tkosciolek commented:
Great!
Disopred seems to work fine. I still haven't tested memsat, will do in an hour or so. But memsat is low priority.

On Feb 27, 2017, at 13:23, Stefan Janssen <notifications@github.com<mailto:notifications@github.com>> wrote:


  *   I changed the message
  *   wrapper is now renamed into run_psipred.pl
  *   I added the option -o to specify an output directory. By default it will be ./psitmpXXX
  *   the *.mtx is now part of the regular output.

  *   I did not add the ability to print the readme file, since it is quit unusual to do something like this. It would be more appropriate to convert it into a man file, but I don't know how to do it.

P.S. do you have further comments on disopred and memsat-svm?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/biocore/conda-recipes/pull/90#issuecomment-282857842>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AGZ0Vhh9LaVLO4vgdCIS_aiw_OHCEIrdks5rgz6kgaJpZM4MKchQ>.
